### DISCUSSION TITLE
Is this a kind of degrade in Genie5?

### DISCUSSION THREAD

**Discussion by onoke123** [GenieFramework/Genie.jl#701] [General]
I compared juliav1.7/v1.10/Genie4/Genie5 with a mutual dependence module model. Then, only Genie5 could not handle it. I must ask you, is this intended design? If it was, will Genie6 hire the same model? Indeed, this change G4 to G5 was a kind of degrade for me.
mutual dependence module model example
```
[a.jl]
module a
        using ..b               <-- julia v1.7/1.10
        using b                  <-- G4
        using GTest.b      <-- G5. 'GTest' is an app name

[b.jl]
module b
        using ..a               <-- julia v1.7/1.10
        using a                 <-- G4
        using GTest.a.    <-- G5 

```

**Comment by essenciary** [comment]
In order to use a module that is not provided by a package in Julia you first need to `include(...)` its file. Genie handles the code loading (the including) of the app files. 

Thus: 
* Julia changed how code loading works which broke the code loading in Genie v4 (well, actually Genie was relying on some "private" Julia APIs which were removed). This required the release of the (breaking) Genie 5 to fix it - and code loading was the most important breaking change. Code loading can no longer work as in v4 due to how code loading works in Julia now. 
* you can always opt-in to `include` the files yourself. However beware that this might break things, like picking up changes in the files (Genie automatically uses `Revise.jl` to reload code changes, and this is done via code loading).

**Reply by onoke123** [reply]
Thank you for your reply, essenciary.
According to your response, GenieV5 loading system is going to be take over to V6, is't it? And below points that I found in V5 are as well?

1. only "resources" dir is autoloading directory
2. .autoload file effects only in "resources" directory

Execute include() before defining "using" is not bad, in fact Genie.jl is hiring this style a lot and it is good reference, but need to beaware its namespace tree. I bet V4 was greater than V5. :p

**Reply by essenciary** [reply]
* There are no changes to code loading planned for v6, so yes, it will stay the same. 
* v5 is cleaner as v4 was exploiting an unofficial feature of LOAD_PATH that was removed by Julia. Also in v5 all the app's resources are namespaced, vs spilling into `Main`. 
* autoloading also works for `config` folder, `lib` etc

**Reply by onoke123** [reply]
Thank you again, essenciary.
Looks like I have to change my app codes in order to your policy.

This is the last question in this discussion, about "autoloading also ....".
 I tried with "lib" dir, but did not work. I mean I got an error "UndefVarError: 'a' not defined" at "a" module that was called in routes.jl, I believe because the module could not be loaded from "lib" dir. Succeeded in "resource" dir. Looked at Loader.jl, only load_resources() is called by others. That why I said only "resources" dir works with autoloading.
 How can I use or declare a module  in routes.jl the module if it works?
"lib" was localted in the same level with "resources".
```
[routes.jl]
using G5Test.a        <-- succeed if module "a" were in "resources", error if it were in "lib"
```

**Reply by essenciary** [reply]
We have a migration guide @PGimenez can you guide @onoke123 please?

**Reply by PGimenez** [reply]
Yes, the guide can be found here

https://learn.genieframework.com/docs/reference/workflow/migrating-genie-4-to-5

**Reply by onoke123** [reply]
Thanks Pere,Adrian, but I read this doc million times. :P
Then gave up migration, started porting. I mean create new app with Genie5, then put my apps into them. Meanwhile I found only "resources" dir worked with autoloading. In Genie site says "lib" also is, maybe it does not be updated yet, and this migration doc says "plugins", "libs" ... I tried them all, the consequence was the both, autoloading and .autoload file, worked fine only "resources" dir.
I just wanna know it is intended and designed or a kind of bug. If it is the restriction, I just follow it. If it is a bug, I am going to try to fix it after releasing our app v1, because I have to focus on my own app so far.
And if you insist other dir,"lib"..., work fine, let me know how to make it, please.

**Comment by onoke123** [comment]
You may say 'use include()', but I am wondering is it correct action or not.
I mean, Genie5 does never kick '__init__()' as far as not using include(). I guess Genie takes a late binding phylosophy and thinks unnecessary it in a web app. But sometimes it is expected and even julia does it. 
I think this is a kind of bug or degrade, do not you think so?
